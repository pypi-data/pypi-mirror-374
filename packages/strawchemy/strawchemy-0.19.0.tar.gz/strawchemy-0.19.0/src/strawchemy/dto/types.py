"""DTO domain types."""

from __future__ import annotations

import dataclasses
from dataclasses import dataclass, field
from enum import Enum
from typing import TYPE_CHECKING, Any, Literal, Optional, Union, final, get_type_hints

from typing_extensions import TypeAlias, override

from strawchemy.utils import get_annotations

if TYPE_CHECKING:
    from collections.abc import Callable, Mapping


__all__ = ("DTOAuto", "DTOConfig", "DTOFieldConfig", "DTOMissing", "ExcludeFields", "IncludeFields", "Purpose")

DTOScope: TypeAlias = Literal["global", "dto"]


IncludeFields: TypeAlias = Union[list[str], set[str], Literal["all"]]
ExcludeFields: TypeAlias = Union[list[str], set[str]]


@final
class DTOMissing:
    """A sentinel type to detect if a parameter is supplied or not when.

    constructing pydantic FieldInfo.
    """


@final
class DTOAuto: ...


@final
class DTOSkip: ...


@final
class DTOUnset:
    @override
    def __str__(self) -> str:
        return ""

    @override
    def __repr__(self) -> str:
        return "DTOUnset"

    def __bool__(self) -> bool:
        return False


class Purpose(str, Enum):
    """For identifying the purpose of a DTO to the factory.

    The factory will exclude fields marked as private or read-only on the domain model depending
    on the purpose of the DTO.

    Example:
    ```python
    ReadDTO = dto.factory("AuthorReadDTO", Author, purpose=dto.Purpose.READ)
    ```
    """

    READ = "read"
    """To mark a DTO that is to be used to serialize data returned to
    clients."""
    WRITE = "write"
    """To mark a DTO that is to deserialize and validate data provided by
    clients."""
    COMPLETE = "complete"
    """To mark a DTO that is to deserialize and validate data provided by
    clients. Fields marked as TO_COMPLETE must not be null."""


@dataclass
class PurposeConfig:
    """Mark the field as read-only, or private."""

    type_override: Optional[Any] = DTOMissing
    validator: Optional[Callable[[Any], Any]] = None
    """Single argument callables that are defined on the DTO as validators for the field."""
    alias: Optional[str] = None
    """Customize name of generated DTO field."""
    partial: Optional[bool] = None


@dataclass
class DTOFieldConfig:
    """For configuring DTO behavior on SQLAlchemy model fields."""

    purposes: set[Purpose] = field(default_factory=lambda: {Purpose.READ, Purpose.WRITE})
    default_config: PurposeConfig = field(default_factory=PurposeConfig)
    configs: dict[Purpose, PurposeConfig] = field(default_factory=dict)

    def purpose_config(self, dto_config: DTOConfig) -> PurposeConfig:
        return self.configs.get(dto_config.purpose, self.default_config)


@dataclass
class DTOConfig:
    """Control the generated DTO.

    This class holds configuration settings that influence how a Data Transfer
    Object (DTO) is generated by the DTO factory. It allows customization of
    field inclusion/exclusion, optionality, type hints, and field aliasing based
    on the intended purpose (read, write, etc.) of the DTO.

    Attributes:
        purpose: Configure the DTO for "read", "write", or "complete" operations.
            Determines which fields from the source model are included based on
            their `DTOFieldConfig`.
        include: Explicitly include fields from the source model in the generated
            DTO. Can be a list or set of field names, or the literal "all" to
            include all fields not explicitly excluded. Defaults to an empty set.
        exclude: Explicitly exclude fields from the source model. Can be a list
            or set of field names. Defaults to an empty set. Setting this
            implicitly sets `include` to "all".
        partial: If True, makes all fields in the generated DTO optional.
            Defaults to None.
        partial_default: The default value assigned to fields when `partial` is
            True and the field is not provided. Defaults to None.
        unset_sentinel: A sentinel object used to represent fields that are not
            set, particularly useful when distinguishing between a field explicitly
            set to `None` and a field that was not provided at all. Defaults to
            `DTO_UNSET`.
        type_overrides: A mapping to override the type annotations for specific
            fields in the generated DTO. Keys can be field names or types,
            values are the overriding types. Defaults to an empty dict.
        annotation_overrides: A dictionary to directly set or override the type
            annotations for specific fields by name. Defaults to an empty dict.
        aliases: A mapping of source model field names to their desired names
            (aliases) in the generated DTO. Defaults to an empty dict. Mutually
            exclusive with `alias_generator`.
        alias_generator: A callable that accepts a field name and returns its
            alias for the generated DTO. Defaults to None. Mutually exclusive
            with `aliases`.

    Raises:
        ValueError: If both `aliases` and `alias_generator` are provided, or
            if `exclude` is set while `include` is also set to a specific list/set
            (i.e., not "all" or empty).
    """

    purpose: Purpose
    """Configure the DTO for "read" or "write" operations."""
    include: IncludeFields = field(default_factory=set)
    """Explicitly include fields from the generated DTO."""
    exclude: ExcludeFields = field(default_factory=set)
    """Explicitly exclude fields from the generated DTO. Implies `include="all"`."""
    partial: Optional[bool] = None
    """Make all field optional."""
    partial_default: Any = None
    unset_sentinel: Any = DTOUnset
    type_overrides: Mapping[Any, Any] = field(default_factory=dict)
    annotation_overrides: dict[str, Any] = field(default_factory=dict)
    aliases: Mapping[str, str] = field(default_factory=dict)
    exclude_defaults: bool = False
    alias_generator: Optional[Callable[[str], str]] = None
    scope: Optional[DTOScope] = None
    exclude_from_scope: bool = False
    tags: set[str] = field(default_factory=set)

    def __post_init__(self) -> None:
        if self.aliases and self.alias_generator is not None:
            msg = "You must set `aliases` or `alias_generator`, not both"
            raise ValueError(msg)
        if self.include and self.include != "all" and self.exclude:
            msg = "When using `exclude` you must set `include='all' or leave it unset`"
            raise ValueError(msg)
        if self.exclude:
            self.include = "all"

    def copy_with(
        self,
        purpose: Union[Purpose, type[DTOUnset]] = DTOUnset,
        include: Optional[IncludeFields] = None,
        exclude: Optional[ExcludeFields] = None,
        partial: Union[bool, None, type[DTOUnset]] = DTOUnset,
        unset_sentinel: Union[Any, type[DTOUnset]] = DTOUnset,
        type_overrides: Union[Mapping[Any, Any], type[DTOUnset]] = DTOUnset,
        annotation_overrides: Union[dict[str, Any], type[DTOUnset]] = DTOUnset,
        aliases: Union[Mapping[str, str], type[DTOUnset]] = DTOUnset,
        exclude_defaults: Union[bool, type[DTOUnset]] = DTOUnset,
        alias_generator: Union[Callable[[str], str], type[DTOUnset]] = DTOUnset,
        partial_default: Union[Any, type[DTOUnset]] = DTOUnset,
        scope: Union[DTOScope, type[DTOUnset]] = DTOUnset,
        exclude_from_scope: Union[bool, type[DTOUnset]] = DTOUnset,
        tags: Union[set[str], type[DTOUnset]] = DTOUnset,
    ) -> DTOConfig:
        """Create a copy of the DTOConfig with the specified changes."""
        if include is None and exclude is None:
            include, exclude = self.include, self.exclude
        else:
            include = include or []
            exclude = exclude or []
        return DTOConfig(
            include=include,
            exclude=exclude,
            purpose=self.purpose if purpose is DTOUnset else purpose,
            partial=self.partial if partial is DTOUnset else partial,
            unset_sentinel=self.unset_sentinel if unset_sentinel is DTOUnset else unset_sentinel,
            type_overrides=self.type_overrides if type_overrides is DTOUnset else type_overrides,
            annotation_overrides=self.annotation_overrides
            if annotation_overrides is DTOUnset
            else annotation_overrides,
            aliases=self.aliases if aliases is DTOUnset else aliases,
            exclude_defaults=self.exclude_defaults if exclude_defaults is DTOUnset else exclude_defaults,
            alias_generator=self.alias_generator if alias_generator is DTOUnset else alias_generator,
            partial_default=self.partial_default if partial_default is DTOUnset else partial_default,
            scope=self.scope if scope is DTOUnset else scope,
            exclude_from_scope=self.exclude_from_scope if exclude_from_scope is DTOUnset else exclude_from_scope,
            tags=self.tags if tags is DTOUnset else tags,
        )

    def with_base_annotations(self, base: type[Any]) -> DTOConfig:
        """Merge type annotations from a base class into this DTOConfig.

        Args:
            base: The base class to extract type annotations from

        Returns:
            A new DTOConfig instance with:
            - Type annotations from the base class merged into annotation_overrides
            - Updated include set to include all fields if exclude is specified or include was "all"

        The method handles two cases:
        1. When include is "all" or exclude is specified: All fields from the base class are included
        2. When specific fields are included: Only those fields are added to the include set
        """
        include: set[str] = set(self.include) if self.include != "all" else set()
        include_all = self.include == "all" or self.exclude
        annotation_overrides: dict[str, Any] = self.annotation_overrides
        try:
            base_annotations = get_type_hints(base, include_extras=True)
        except NameError:
            base_annotations = get_annotations(base)
        for name, annotation in base_annotations.items():
            if not include_all:
                include.add(name)
            annotation_overrides[name] = annotation
        return dataclasses.replace(
            self,
            include="all" if include_all else include,
            annotation_overrides=annotation_overrides,
        )

    def alias(self, name: str) -> Optional[str]:
        if self.aliases:
            return self.aliases.get(name)
        if self.alias_generator is not None:
            return self.alias_generator(name)
        return None
