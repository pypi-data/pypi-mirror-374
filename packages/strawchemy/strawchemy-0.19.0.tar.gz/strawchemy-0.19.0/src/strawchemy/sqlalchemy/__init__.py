"""SQLAlchemy integration for Strawchemy.

This package provides SQLAlchemy-based repository implementations for use with
Strawberry GraphQL.
"""

from __future__ import annotations

from .repository import SQLAlchemyGraphQLAsyncRepository, SQLAlchemyGraphQLRepository, SQLAlchemyGraphQLSyncRepository

__all__ = ("SQLAlchemyGraphQLAsyncRepository", "SQLAlchemyGraphQLRepository", "SQLAlchemyGraphQLSyncRepository")
