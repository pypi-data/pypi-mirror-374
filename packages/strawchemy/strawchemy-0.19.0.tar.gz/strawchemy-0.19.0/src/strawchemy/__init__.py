"""Custom DTO implementation."""

from __future__ import annotations

from .config.base import StrawchemyConfig
from .mapper import Strawchemy
from .sqlalchemy.hook import QueryHook
from .strawberry import ModelInstance
from .strawberry.mutation.input import Input
from .strawberry.mutation.types import (
    ErrorType,
    RequiredToManyUpdateInput,
    RequiredToOneInput,
    ToManyCreateInput,
    ToManyUpdateInput,
    ToOneInput,
    ValidationErrorType,
)
from .strawberry.repository import StrawchemyAsyncRepository, StrawchemySyncRepository
from .validation.base import InputValidationError

__all__ = (
    "ErrorType",
    "Input",
    "InputValidationError",
    "ModelInstance",
    "QueryHook",
    "RequiredToManyUpdateInput",
    "RequiredToOneInput",
    "Strawchemy",
    "StrawchemyAsyncRepository",
    "StrawchemyConfig",
    "StrawchemySyncRepository",
    "ToManyCreateInput",
    "ToManyUpdateInput",
    "ToOneInput",
    "ValidationErrorType",
)
