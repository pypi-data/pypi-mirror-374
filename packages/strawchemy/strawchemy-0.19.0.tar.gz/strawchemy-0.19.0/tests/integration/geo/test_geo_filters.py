from __future__ import annotations

import pytest

pytest.importorskip("geoalchemy2", reason="geoalchemy2 is not installed")


from typing import TYPE_CHECKING, Any

from sqlalchemy import Executable, Insert, MetaData, insert, text
from tests.integration.fixtures import QueryTracker
from tests.integration.geo.types import mysql as mysql_types
from tests.integration.geo.types import postgres as postgres_types
from tests.integration.utils import to_graphql_representation
from tests.typing import AnyQueryExecutor
from tests.utils import maybe_async

from .models import GeoModel, geo_metadata

if TYPE_CHECKING:
    from pytest_databases.docker.postgres import PostgresService
    from strawchemy.typing import SupportedDialect

    from syrupy.assertion import SnapshotAssertion
    from tests.integration.typing import RawRecordData


pytestmark = [pytest.mark.integration, pytest.mark.geo]


@pytest.fixture
def metadata() -> MetaData:
    return geo_metadata


@pytest.fixture
def postgres_database_service(postgis_service: PostgresService) -> PostgresService:
    return postgis_service


@pytest.fixture
def before_create_all_statements(dialect: SupportedDialect) -> list[Executable]:
    if dialect == "postgresql":
        return [text("CREATE EXTENSION IF NOT EXISTS postgis")]
    return []


@pytest.fixture
def async_query(dialect: SupportedDialect) -> type[Any]:
    if dialect == "postgresql":
        return postgres_types.AsyncGeoQuery
    if dialect == "mysql":
        return mysql_types.AsyncGeoQuery
    pytest.skip(f"Geo tests can't be run on this dialect: {dialect}")


@pytest.fixture
def sync_query(dialect: SupportedDialect) -> type[Any]:
    if dialect == "postgresql":
        return postgres_types.SyncGeoQuery
    if dialect == "mysql":
        return mysql_types.SyncGeoQuery
    pytest.skip(f"Geo tests can't be run on this dialect: {dialect}")


@pytest.fixture
def seed_insert_statements(raw_geo: RawRecordData) -> list[Insert]:
    return [insert(GeoModel).values(raw_geo)]


@pytest.mark.snapshot
async def test_no_filtering(
    any_query: AnyQueryExecutor, raw_geo: RawRecordData, query_tracker: QueryTracker, sql_snapshot: SnapshotAssertion
) -> None:
    """Test that querying without filters returns all records."""
    result = await maybe_async(any_query("{ geoField { id } }"))
    assert not result.errors
    assert result.data
    assert len(result.data["geoField"]) == len(raw_geo)
    # Verify SQL query
    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.parametrize(
    ("field_name", "geometry", "expected_ids"),
    [
        pytest.param("polygon", {"type": "Point", "coordinates": [0.5, 0.5]}, [0], id="point-within-polygon"),
        pytest.param("multiPolygon", {"type": "Point", "coordinates": [2.5, 2.5]}, [0], id="point-within-multipolygon"),
        pytest.param("geometry", {"type": "Point", "coordinates": [5, 5]}, [0], id="point-equals-geometry-point"),
    ],
)
@pytest.mark.snapshot
async def test_contains_geometry(
    field_name: str,
    geometry: dict[str, Any],
    expected_ids: list[int],
    any_query: AnyQueryExecutor,
    raw_geo: RawRecordData,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
) -> None:
    """Test the contains_geometry filter.

    This tests that a geometry contains the provided geometry.
    For example, a polygon contains a point if the point is inside the polygon.
    """
    query = f"""
            {{
                geoField(filter: {{ {field_name}: {{ containsGeometry: {to_graphql_representation(geometry, "input")} }} }}) {{
                    id
                    {field_name}
                }}
            }}
    """
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data
    assert len(result.data["geoField"]) == len(expected_ids)
    for i, expected_id in enumerate(expected_ids):
        assert result.data["geoField"][i]["id"] == raw_geo[expected_id]["id"]
    # Verify SQL query
    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.parametrize(
    ("field_name", "geometry", "expected_ids"),
    [
        pytest.param(
            "point",
            {"type": "Polygon", "coordinates": [[[0, 0], [0, 2], [2, 2], [2, 0], [0, 0]]]},
            [0],
            id="point-within-polygon",
        ),
        pytest.param(
            "lineString",
            {"type": "Polygon", "coordinates": [[[0, 0], [0, 3], [3, 3], [3, 0], [0, 0]]]},
            [0],
            id="linestring-within-polygon",
        ),
        pytest.param(
            "multiPoint",
            {
                "coordinates": [
                    [
                        [3.8530909369223423, 3.5077205177229587],
                        [-2.126498556883888, 3.5077205177229587],
                        [-2.126498556883888, -0.8070832671228061],
                        [3.8530909369223423, -0.8070832671228061],
                        [3.8530909369223423, 3.5077205177229587],
                    ]
                ],
                "type": "Polygon",
            },
            [0],
            id="multipoint-within-polygon",
        ),
    ],
)
@pytest.mark.snapshot
async def test_within_geometry(
    field_name: str,
    geometry: dict[str, Any],
    expected_ids: list[int],
    any_query: AnyQueryExecutor,
    raw_geo: RawRecordData,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
) -> None:
    """Test the within_geometry filter.

    This tests that a geometry is within the provided geometry.
    For example, a point is within a polygon if the point is inside the polygon.
    """
    query = f"""
            {{
                geoField(filter: {{ {field_name}: {{ withinGeometry: {to_graphql_representation(geometry, "input")} }} }}) {{
                    id
                    {field_name}
                }}
            }}
    """
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data
    assert len(result.data["geoField"]) == len(expected_ids)
    for i, expected_id in enumerate(expected_ids):
        assert result.data["geoField"][i]["id"] == raw_geo[expected_id]["id"]
    # Verify SQL query
    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.snapshot
async def test_is_null(
    any_query: AnyQueryExecutor, raw_geo: RawRecordData, query_tracker: QueryTracker, sql_snapshot: SnapshotAssertion
) -> None:
    """Test the isNull filter for geometry fields."""
    query = """
            {
                geoField(filter: { point: { isNull: true } }) {
                    id
                    point
                }
            }
    """
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data
    assert len(result.data["geoField"]) == 1
    assert result.data["geoField"][0]["id"] == raw_geo[1]["id"]
    assert result.data["geoField"][0]["point"] is None
    # Verify SQL query
    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot
