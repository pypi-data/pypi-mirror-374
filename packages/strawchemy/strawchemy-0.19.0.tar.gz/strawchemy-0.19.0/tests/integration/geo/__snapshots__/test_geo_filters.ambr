# serializer version: 1
# name: test_contains_geometry[session-tracked-async-point-equals-geometry-point-asyncmy_engine]
  '''
  SELECT ST_AsBinary(geos_fields.geometry) AS geometry,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.geometry, ST_GeomFromGeoJSON(%s))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_contains_geometry[session-tracked-async-point-equals-geometry-point-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.geometry) AS geometry,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.geometry, ST_GeomFromGeoJSON($1::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_contains_geometry[session-tracked-async-point-equals-geometry-point-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.geometry) AS geometry,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.geometry, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_contains_geometry[session-tracked-async-point-within-geometry-polygon-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.geometry) AS geometry,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.geometry, ST_GeomFromGeoJSON($1::VARCHAR))
  '''
# ---
# name: test_contains_geometry[session-tracked-async-point-within-geometry-polygon-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.geometry) AS geometry,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.geometry, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
# name: test_contains_geometry[session-tracked-async-point-within-multipolygon-asyncmy_engine]
  '''
  SELECT ST_AsBinary(geos_fields.multi_polygon) AS multi_polygon,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.multi_polygon, ST_GeomFromGeoJSON(%s))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_contains_geometry[session-tracked-async-point-within-multipolygon-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.multi_polygon) AS multi_polygon,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.multi_polygon, ST_GeomFromGeoJSON($1::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_contains_geometry[session-tracked-async-point-within-multipolygon-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.multi_polygon) AS multi_polygon,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.multi_polygon, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_contains_geometry[session-tracked-async-point-within-polygon-asyncmy_engine]
  '''
  SELECT ST_AsBinary(geos_fields.polygon) AS
  POLYGON,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.polygon, ST_GeomFromGeoJSON(%s))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_contains_geometry[session-tracked-async-point-within-polygon-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.polygon) AS
  POLYGON,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.polygon, ST_GeomFromGeoJSON($1::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_contains_geometry[session-tracked-async-point-within-polygon-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.polygon) AS
  POLYGON,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.polygon, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_contains_geometry[session-tracked-sync-point-equals-geometry-point-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.geometry) AS geometry,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.geometry, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_contains_geometry[session-tracked-sync-point-within-geometry-polygon-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.geometry) AS geometry,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.geometry, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
# name: test_contains_geometry[session-tracked-sync-point-within-multipolygon-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.multi_polygon) AS multi_polygon,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.multi_polygon, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_contains_geometry[session-tracked-sync-point-within-polygon-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.polygon) AS
  POLYGON,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.polygon, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_is_null[session-tracked-async-asyncmy_engine]
  '''
  SELECT ST_AsBinary(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE geos_fields.point IS NULL
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_is_null[session-tracked-async-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE geos_fields.point IS NULL
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_is_null[session-tracked-async-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE geos_fields.point IS NULL
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_is_null[session-tracked-sync-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE geos_fields.point IS NULL
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_no_filtering[session-tracked-async-asyncmy_engine]
  '''
  SELECT geos_fields.id
    FROM geos_fields AS geos_fields
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_no_filtering[session-tracked-async-asyncpg_engine]
  '''
  SELECT geos_fields.id
    FROM geos_fields AS geos_fields
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_no_filtering[session-tracked-async-psycopg_async_engine]
  '''
  SELECT geos_fields.id
    FROM geos_fields AS geos_fields
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_no_filtering[session-tracked-sync-psycopg_engine]
  '''
  SELECT geos_fields.id
    FROM geos_fields AS geos_fields
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_within_geometry[session-tracked-async-linestring-within-polygon-asyncmy_engine]
  '''
  SELECT ST_AsBinary(geos_fields.line_string) AS line_string,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.line_string, ST_GeomFromGeoJSON(%s))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_within_geometry[session-tracked-async-linestring-within-polygon-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.line_string) AS line_string,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.line_string, ST_GeomFromGeoJSON($1::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_within_geometry[session-tracked-async-linestring-within-polygon-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.line_string) AS line_string,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.line_string, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_within_geometry[session-tracked-async-multipoint-within-polygon-asyncmy_engine]
  '''
  SELECT ST_AsBinary(geos_fields.multi_point) AS multi_point,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.multi_point, ST_GeomFromGeoJSON(%s))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_within_geometry[session-tracked-async-multipoint-within-polygon-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.multi_point) AS multi_point,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.multi_point, ST_GeomFromGeoJSON($1::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_within_geometry[session-tracked-async-multipoint-within-polygon-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.multi_point) AS multi_point,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.multi_point, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_within_geometry[session-tracked-async-point-within-polygon-asyncmy_engine]
  '''
  SELECT ST_AsBinary(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.point, ST_GeomFromGeoJSON(%s))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_within_geometry[session-tracked-async-point-within-polygon-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.point, ST_GeomFromGeoJSON($1::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_within_geometry[session-tracked-async-point-within-polygon-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.point, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_within_geometry[session-tracked-sync-linestring-within-polygon-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.line_string) AS line_string,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.line_string, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_within_geometry[session-tracked-sync-multipoint-within-polygon-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.multi_point) AS multi_point,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.multi_point, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
# name: test_within_geometry[session-tracked-sync-point-within-polygon-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.point, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
   ORDER BY geos_fields.id ASC
  '''
# ---
